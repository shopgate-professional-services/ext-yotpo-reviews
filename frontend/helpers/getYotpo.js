/**
 * Mounts yotpo js widget script.
 * @param {string} config Yotpo app key.
 */
function mountScript(config) {
  const script = document.createElement('script');
  script.type = 'text/javascript';
  script.async = true;
  script.src = `//staticw2.yotpo.com/${config}/widget.js`;
  script.id = 'yotpoWidgetScript';
  const parent = document.getElementsByTagName('head')[0];
  parent.appendChild(script);
}

/**
 * Tries to get Yotpo from window Object. If not yet available, tries again.
 * @param {function} resolve Promise.resolve function
 */
function tryYotpo(resolve) {
  if (window.Yotpo) {
    resolve(window.Yotpo);
    return;
  }
  setTimeout(() => {
    tryYotpo(resolve);
  }, 100);
}

/**
 * Injects Yotpo script and returns promise of the Yotpo Object.
 * @returns {Promise<Object>}
 * @param {string} config Yotpo app key.
 */
function yotpoPromise(config) {
  mountScript(config);
  return new Promise((resolve) => {
    tryYotpo(resolve);
  });
}

/**
 * @type {Promise}
 */
let promise;

/**
 * Gets the Yotpo Object
 * @returns {Promise}
 * @param {string} config Yotpo app key.
 */
function getYotpo(config) {
  if (!promise) {
    promise = yotpoPromise(config);
  }
  return promise;
}

export default getYotpo;
