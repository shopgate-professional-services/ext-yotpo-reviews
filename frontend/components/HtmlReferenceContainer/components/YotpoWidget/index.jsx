import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './style';
import getYotpo from '../../../../helpers/getYotpo';

/**
 * Mounts yotpo script.
 */
class YotpoScript extends Component {
  static propTypes = {
    config: PropTypes.shape().isRequired,
    productId: PropTypes.string.isRequired,
    product: PropTypes.shape(),
  };

  static defaultProps = {
    product: null,
  }

  /**
   * Constructs
   * @param {Objects} props props
   */
  constructor(props) {
    super(props);
    this.yotpoMountTimeout = undefined;
  }

  /**
   * Mounts Yotpo script.
   */
  componentDidMount() {
    if (!window.yotpo) {
      getYotpo(this.props.config.yotpoAppKey);
      return;
    }
    window.yotpo.initialized = false;
    window.yotpo.clean();
    this.yotpoMountTimeout = setTimeout(() => {
      window.yotpo.initWidgets();
      this.yotpoMountTimeout = undefined;
    }, 500);
  }

  /**
   * Cleanup timeout request
   */
  componentWillUnmount() {
    clearTimeout(this.yotpoMountTimeout);
  }

  /**
   * Renders.
   * @return {JSX}
   */
  render() {
    if (!this.props.product) {
      return null;
    }
    return (
      <div>
        <div
          className={`yotpo bottomLine ${styles.padding}`}
          data-appkey={`${this.props.config.yotpoAppKey}`}
          data-domain=""
          data-product-id={`${this.props.productId}`}
          data-name={`${this.props.product.name}`}
          data-url={`${this.props.product.productUrl}`}
          data-image-url={`${this.props.product.featuredImageUrl}`}
          data-description=""
        />
        <div
          className={`yotpo yotpo-main-widget ${styles.container} ${styles.padding}`}
          data-product-id={`${this.props.productId}`}
          data-name={`${this.props.product.name}`}
          data-url={`${this.props.product.productUrl}`}
          data-image-url={`${this.props.product.featuredImageUrl}`}
          data-description=""
        />
      </div>
    );
  }
}

export default YotpoScript;
