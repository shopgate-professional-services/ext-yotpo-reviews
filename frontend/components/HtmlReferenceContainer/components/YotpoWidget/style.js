import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { variables } = themeConfig;

const padding = css({
  padding: `0px ${variables.gap.big}px 0px ${variables.gap.big}px`,
}).toString();

const container = css({
  maxWidth: '100%',
  wordWrap: 'break-word',
  wordBreak: 'bread-word',
  ' .socialize-wrapper': {
    display: 'none !important',
  },
}).toString();

export default {
  container,
  padding,
};
