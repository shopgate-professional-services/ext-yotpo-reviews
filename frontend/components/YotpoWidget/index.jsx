import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './style';
import getYotpo from '../../helpers/getYotpo';

/**
 * Mounts yotpo script.
 */
class YotpoScript extends Component {
  static propTypes = {
    config: PropTypes.shape().isRequired,
    productId: PropTypes.string.isRequired,
    product: PropTypes.shape(),
  };

  static defaultProps = {
    product: null,
  }

  /**
   * Mounts Yotpo script.
   */
  componentDidMount() {
    if (!window.yotpo) {
      getYotpo(this.props.config.yotpoAppKey);
      return;
    }
    window.yotpo.initialized = false;
    window.yotpo.clean();
    setTimeout(() => { window.yotpo.initWidgets(); }, 500);
  }

  /**
   * Update for yotpo fetching for different variant
   * @param {Object} prevProps prevProps
   */
  componentDidUpdate(prevProps) {
    if (this.props.productId !== prevProps.productId) {
      window.yotpo.initialized = false;
      window.yotpo.clean();
      // 1500 Additional ms to account for refresh of widget properties and fetching of new widget.
      setTimeout(() => {
        window.yotpo.refreshWidgets();
        window.yotpo.initWidgets();
      }, 2000);
    }
  }

  /**
   * Renders.
   * @return {JSX}
   */
  render() {
    if (!this.props.product) {
      return null;
    }
    return (
      <div>
        <div
          className={`yotpo bottomLine ${styles.padding}`}
          data-appkey={`${this.props.config.yotpoAppKey}`}
          data-domain=""
          data-product-id={`${this.props.productId}`}
          data-name={`${this.props.product.name}`}
          data-url={`${this.props.product.productUrl}`}
          data-image-url={`${this.props.product.featuredImageUrl}`}
          data-description=""
        />
        <div
          className={`yotpo yotpo-main-widget ${styles.container} ${styles.padding}`}
          data-product-id={`${this.props.productId}`}
          data-name={`${this.props.product.name}`}
          data-url={`${this.props.product.productUrl}`}
          data-image-url={`${this.props.product.featuredImageUrl}`}
          data-description=""
        />
      </div>
    );
  }
}

export default YotpoScript;
