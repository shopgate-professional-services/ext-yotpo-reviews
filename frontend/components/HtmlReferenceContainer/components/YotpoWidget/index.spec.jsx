import React from 'react';
import { mount } from 'enzyme';
import YotpoWidget from './index';
import getYotpo from '../../../../helpers/getYotpo';

jest.mock('../../../../helpers/getYotpo');
const noProduct = null;
const product = {
  name: 'item',
  productUrl: 'wwww.123.com',
  featuredImageUrl: 'www.123456.com',
};
describe('<YotpoWidget />', () => {
  it('should not render if there is product props', () => {
    const wrapper = mount((
      <YotpoWidget product={noProduct} config={{ yotpoAppKey: null }} productId="2" />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').exists()).toBe(false);
  });
  it('should render yotpo divs when a product is passed', () => {
    const wrapper = mount((
      <YotpoWidget product={product} config={{ yotpoAppKey: null }} productId="2" />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').exists()).toBe(true);
  });
  it('should call mountScript if no window.yotpo', () => {
    const wrapper = mount((
      <YotpoWidget product={product} config={{ yotpoAppKey: null }} productId="2" />
    ));
    const renderedInstance = wrapper.instance();
    renderedInstance.componentDidMount();
    expect(getYotpo).toHaveBeenCalled();
  });
  it('should call clean from window.yotpo', () => {
    window.yotpo = {
      initialzied: false,
      clean: jest.fn(),
      initWidgets: jest.fn(),
      refreshWidgets: jest.fn(),
    };
    const wrapper = mount((
      <YotpoWidget product={product} config={{ yotpoAppKey: null }} productId="2" />
    ));
    const renderedInstance = wrapper.instance();
    renderedInstance.componentDidMount();
    expect(window.yotpo.clean).toHaveBeenCalled();
  });
});
