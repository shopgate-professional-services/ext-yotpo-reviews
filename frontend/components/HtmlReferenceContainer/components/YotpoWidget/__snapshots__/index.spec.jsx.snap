// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YotpoWidget /> should not render if there is product props 1`] = `
<YotpoScript
  config={
    Object {
      "yotpoAppKey": null,
    }
  }
  product={null}
  productId="2"
/>
`;

exports[`<YotpoWidget /> should render yotpo divs when a product is passed 1`] = `
<YotpoScript
  config={
    Object {
      "yotpoAppKey": null,
    }
  }
  product={
    Object {
      "featuredImageUrl": "www.123456.com",
      "name": "item",
      "productUrl": "wwww.123.com",
    }
  }
  productId="2"
>
  <div>
    <div
      className="yotpo bottomLine css-1gv7y8"
      data-appkey="null"
      data-description=""
      data-domain=""
      data-image-url="www.123456.com"
      data-name="item"
      data-product-id="2"
      data-url="wwww.123.com"
    />
    <div
      className="yotpo yotpo-main-widget css-12ctqp0 css-1gv7y8"
      data-description=""
      data-image-url="www.123456.com"
      data-name="item"
      data-product-id="2"
      data-url="wwww.123.com"
    />
  </div>
</YotpoScript>
`;
